File: GuardianX-main/guardianx-frontend/src/types/asset.ts
export type AssetType =
  | "SERVER"
  | "WORKSTATION"
  | "WEBSITE"
  | "DOMAIN"
  | "IP_ADDRESS"
  | "API"
  | "CLOUD"
  | "MOBILE"
  | "OTHER";

export type Criticality =
  | "LOW"
  | "MEDIUM"
  | "HIGH"
  | "CRITICAL";

export interface Asset {
  id: number;
  name: string;
  asset_type: AssetType;

  ip_address: string | null;
  domain: string | null;
  operating_system: string | null;
  environment: string | null;
  owner: string | null;
  criticality: Criticality | null;
  description: string | null;

  created_by: number;
  created_at: string;
  updated_at: string;
}

export interface CreateAssetDto {
  name: string;
  asset_type: AssetType;

  ip_address?: string;
  domain?: string;
  operating_system?: string;
  environment?: string;
  owner?: string;
  criticality?: Criticality;
  description?: string;
}

export interface UpdateAssetDto {
  name?: string;
  asset_type?: AssetType;

  ip_address?: string;
  domain?: string;
  operating_system?: string;
  environment?: string;
  owner?: string;
  criticality?: Criticality;
  description?: string;
}

export interface AssetServiceItem {
  port: number;
  protocol: string;
  product: string | null;
  version: string | null;
  cpe: string | null;
  state: string;
}

export interface AssetRecentScan {
  scan_id: number;
  status: string;
  started_at: string | null;
  finished_at: string | null;
  total_findings: number;
}

export interface AssetRecentFinding {
  cve: string | null;
  severity: string;
  title: string;
  status: string;
  recommendation: string | null;
  cvss: number | null;
}

export interface AssetDetails {
  id: number;
  name: string;
  hostname: string | null;
  ip_address: string | null;
  asset_type: AssetType;
  domain: string | null;
  operating_system: string | null;
  environment: string | null;
  owner: string | null;
  criticality: Criticality | null;
  description: string | null;

  created_at: string;
  updated_at: string;

  risk_score: number;
  security_score: number;
  total_findings: number;
  critical: number;
  high: number;
  medium: number;
  low: number;

  attack_surface_score: number;
  internet_facing: boolean;
  open_ports: number[];
  technologies: string[];
  ai_summary: string;

  services: AssetServiceItem[];
  recent_scans: AssetRecentScan[];
  recent_findings: AssetRecentFinding[];
}

File: GuardianX-main/backend/app/schemas/asset.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.core.asset_types import AssetType
from app.core.network import validate_scan_target


def _validate_target_field(value):
    if value is None or not str(value).strip():
        return value

    return validate_scan_target(value)


class AssetCreate(BaseModel):
    name: str = Field(
        min_length=2,
        max_length=150,
    )

    asset_type: AssetType

    ip_address: str | None = None
    domain: str | None = None
    operating_system: str | None = None
    environment: str | None = None
    owner: str | None = None
    criticality: str | None = None
    description: str | None = None

    _validate_target = field_validator(
        "ip_address",
        "domain",
        mode="before",
    )(_validate_target_field)


class AssetUpdate(BaseModel):
    name: str | None = None
    asset_type: AssetType | None = None

    ip_address: str | None = None
    domain: str | None = None
    operating_system: str | None = None
    environment: str | None = None
    owner: str | None = None
    criticality: str | None = None
    description: str | None = None

    _validate_target = field_validator(
        "ip_address",
        "domain",
        mode="before",
    )(_validate_target_field)


class AssetResponse(BaseModel):
    id: int

    name: str
    asset_type: AssetType

    ip_address: str | None
    domain: str | None
    operating_system: str | None
    environment: str | None
    owner: str | None
    criticality: str | None
    description: str | None

    created_by: int
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(
        from_attributes=True,
    )


class AssetServiceItem(BaseModel):
    port: int
    protocol: str
    product: str | None = None
    version: str | None = None
    cpe: str | None = None
    state: str


class AssetRecentScanItem(BaseModel):
    scan_id: int
    status: str
    started_at: datetime | None = None
    finished_at: datetime | None = None
    total_findings: int = 0


class AssetRecentFindingItem(BaseModel):
    cve: str | None = None
    severity: str
    title: str
    status: str
    recommendation: str | None = None
    cvss: float | None = None


class AssetDetailResponse(BaseModel):
    id: int
    name: str
    hostname: str | None = None
    ip_address: str | None = None
    asset_type: AssetType
    domain: str | None = None
    operating_system: str | None = None
    environment: str | None = None
    owner: str | None = None
    criticality: str | None = None
    description: str | None = None
    created_at: datetime
    updated_at: datetime

    risk_score: int = 0
    security_score: int = 0
    total_findings: int = 0
    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0

    attack_surface_score: int = 0
    internet_facing: bool = False
    open_ports: list[int] = Field(default_factory=list)
    technologies: list[str] = Field(default_factory=list)
    ai_summary: str = ""

    services: list[AssetServiceItem] = Field(default_factory=list)
    recent_scans: list[AssetRecentScanItem] = Field(default_factory=list)
    recent_findings: list[AssetRecentFindingItem] = Field(default_factory=list)

    model_config = ConfigDict(
        from_attributes=True,
    )
